File: php/src/Role.php
<?php

declare(strict_types=1);

namespace Anubis;

/**
 * A granted role, which Anubis returns prefixed with the application that
 * defined it: a manifest declaring "clerk" for "billing" yields "billing.clerk".
 *
 * Comparing a returned role against the unprefixed manifest name is the
 * mistake this type exists to make visible.
 */
final readonly class Role implements \Stringable
{
    private function __construct(public string $value)
    {
    }

    public static function of(self|string $role): self
    {
        return $role instanceof self ? $role : new self($role);
    }

    public static function from(string $app, string $name): self
    {
        return new self("{$app}.{$name}");
    }

    /** The application that defined the role, or "" for an unprefixed one. */
    public function app(): string
    {
        $dot = strpos($this->value, '.');
        return $dot === false ? '' : substr($this->value, 0, $dot);
    }

    /** The role as the manifest declared it, without the prefix. */
    public function name(): string
    {
        $dot = strpos($this->value, '.');
        return $dot === false ? $this->value : substr($this->value, $dot + 1);
    }

    public function equals(self|string $other): bool
    {
        return $this->value === self::of($other)->value;
    }

    public function __toString(): string
    {
        return $this->value;
    }
}

File: java/src/main/java/io/github/gsoultan/anubis/Role.java
package io.github.gsoultan.anubis;

import java.util.Objects;

/**
 * A granted role, which Anubis returns prefixed with the application that
 * defined it: a manifest declaring {@code clerk} for {@code billing} yields
 * {@code billing.clerk}.
 *
 * <p>Comparing a returned role against the unprefixed manifest name is the
 * mistake this type exists to make visible.
 */
public final class Role {
    private final String value;

    private Role(String value) {
        this.value = value == null ? "" : value;
    }

    public static Role of(String role) {
        return new Role(role);
    }

    public static Role from(String app, String name) {
        return new Role(app + "." + name);
    }

    public String value() {
        return value;
    }

    /** The application that defined the role, or "" for an unprefixed one. */
    public String app() {
        int dot = value.indexOf('.');
        return dot < 0 ? "" : value.substring(0, dot);
    }

    /** The role as the manifest declared it, without the prefix. */
    public String name() {
        int dot = value.indexOf('.');
        return dot < 0 ? value : value.substring(dot + 1);
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Role other && value.equals(other.value);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(value);
    }

    @Override
    public String toString() {
        return value;
    }
}
